### ISSUE TITLE
Training settings for Val14 benchmark

### ISSUE THREAD

**Issue by lan-lc** [autonomousvision/tuplan_garage#11]
Thanks for open sourcing the great work!

I want to ask about the training details of PDM-open and PDM-Offset. 

- Hardwares: 
    - GPU type and numbers
    - CPU cores numbers
    - Maximum Memory Usage
- Training time
    - How many hours per epoch
    - How many epochs do you used?
- Cache
    - Do you save the features like central lines that you calculated?
        - If yes, what is the size of it and how to save and load it?

Thank you in advance.

**Comment by kashyap7x** [comment]
We used a machine with an NVIDIA RTX 3090 GPU and AMD Ryzen 7950X CPU (16 cores, 64GB memory) for training. We trained the models for 100 epochs with a batch size of 64. The feature cache is automatically generated by the nuPlan devkit's training script before the first training epoch. With only the history, centerline, and pdm-closed trajectory as features, its size should be <1GB. The entire training should take around ~30min.

You can find additional details in [Daniel's thesis.](https://danieldauner.github.io/assets/pdf/Dauner2023Master.pdf)

**Comment by lan-lc** [comment]
Thanks for your reply and sharing the thesis!

According to the thesis, you have ~178k of training frames. I wonder how many frames do you extract from one scenario (15s)?

**Comment by kashyap7x** [comment]
The training set includes data from 70 different scenario types, with up to 4k scenarios per type. The scenario filter is applied to leave a minimum interval of 15 seconds between extracted frames, so we only extract **one training frame per scenario**.

**Comment by lan-lc** [comment]
Thank you for the reply! I have two follow-up questions. 

1. Do you randomly sample each training frame from the 0~7 seconds of a scenario (since the prediction is 8 seconds)?
2. How can I extract multiple frames from a single scenario?

Thanks!

**Comment by kashyap7x** [comment]
Sorry for the earlier confusion!

The nuplan devkit does not provide highly flexible ways to sample data for training as mentioned in your questions 1 and 2, but it is possible to train on frames that are temporally close to each other by setting small (or null) values for `timestamp_threshold_s` in the scenario filter.

The training data for our planners is actually filtered from the training log files with this [scenario filter](https://github.com/autonomousvision/tuplan_garage/blob/main/tuplan_garage/planning/script/config/common/scenario_filter/train150k_split.yaml). Here, we did not set the `timestamp_threshold_s` value, which means that our training set may already be using frames that are closer than 15 seconds apart. My earlier answer referred to the [validation set](https://github.com/autonomousvision/tuplan_garage/blob/main/tuplan_garage/planning/script/config/common/scenario_filter/val14_split.yaml), where we use `timestamp_threshold_s=15`.

These other issues may help, please have a look:

https://github.com/motional/nuplan-devkit/issues/269#issuecomment-1512693196

https://github.com/motional/nuplan-devkit/issues/203#issuecomment-1378522525

Hope that helps.

**Comment by lan-lc** [comment]
Hello! Thank you for the detailed explanation. I have a question for clarification.

It seems like for each training frame, it is the first frame of a sampled scenario. Which means target roadblock is always 15s ahead. Is that correct?

If so, this would imply that the PDM-open network is never trained on frames that are near the target. As a result, it's understandable why PDM-open performs poorly in close-loop evaluations, since it has to deal with frames that are close to the target roadblock (which is out of training distribution). 

In addition, when I evaluated PDM-closed on closed-loop non-react, it takes about 14 hours. Is there way to make it faster? (Such as avoiding generating videos).

Thanks in advance!

**Comment by DanielDauner** [comment]
Hi @lan-lc,

Yes, PDM-Open is trained always on the first frame of a scenario. However, the final roadblock is far away and practically not reachable within a scenario's ~15-second closed-loop simulation. The centerline is updated every iteration in a closed-loop and restricted to 120m to avoid distribution shifts. I recommend viewing some scenarios in the nuboard, where the route roadblocks and simulations are visualized.

The `worker=ray_distributed` mode is generally the fastest and parallelizes each scenario simulation across logical cores in your machine. With an `AMD Ryzen 9 7950X` CPU, the Val14 simulation of PDM-Closed takes about 20-30 min. Thus, 14 hours sounds odd. Which argument for `worker` are you currently using? For example, `worker=sequential` sequentially simulates the scenarios on your machine, which would be pretty slow.

**Comment by lan-lc** [comment]
Hi @DanielDauner,

Thanks for your reply!

I didn't notice that the target roadblock is practically not reachable. I initially believed that the target roadblock is the roadblock where the human demonstrations end up at the end of the scenario. 

I understand that centerline is updated every iteration in a closed-looped. However, if the target roadblock is still the same, it will still be an out-of-distribution input for PDM-open right?
In addition, when you mentioned "restricted to 120m," is that referring to the maximum length allowed for the input centerline?   I wonder what the centerline input will be like if the target roadblock is super close (less than 120 m).

For evaluation, by removing "number_of_gpus_allocated_per_simulation=1" and "number_of_gpus_allocated_per_simulation=32" from my command and adding "worker=ray_distributed", PDM-Closed closed-loop NR can be done in 40 mins!  Thanks for the advice!

**Comment by DanielDauner** [comment]
Hi @lan-lc,

First, I am happy to hear that the simulation is now running in a reasonable time. 

Secondly, all learned models are somewhat out-of-distribution (OOD) during closed-loop simulation. The planner steers the car (instead of the human driver), whereas only open-loop scenarios are encountered during training. This is an inherent problem in classical behavior cloning. 

I don't think the centerline and target roadblocks are problems during closed-loop simulation. The centerline is 120m long, with poses at 1m intervals. Given the far distance of the last roadblock in the route sequence, the updated centerline will almost surely have the same length again of 120m. 

I have not encountered a scenario yet where the last roadblock is within reach. It can happen that no 120m centerline is available, e.g. because the graph search couldn't find a solution. In that case, the final pose is repeated since we clip the values when interpolating on the centerline (see [here](https://github.com/autonomousvision/tuplan_garage/blob/e4d510d98cc0018e6e0a9f12ba6ec6f32ea93faf/tuplan_garage/planning/simulation/planner/pdm_planner/utils/pdm_path.py#L72C66-L72C66)). However, this can also occur during training; thus, this it's not directly OOD.

**Comment by DanielDauner** [comment]
I am closing this issue for now. Feel free to re-open if further questions arise.
